### DISCUSSION TITLE
init properties: Do they also work for normal fields?

### DISCUSSION THREAD

**Discussion by msedi** [dotnet/csharplang#4057] [Q&A] [answered]
As the new init-only properties will come I came into a situation today where the question came up if they also work for fields?
So my question is if this would work? For hi-perf computing it's not always helpful to have getters and setters, so I think the situtation as shown below is quite common:

``` CS
struct Point
{
  int init X;
  int init Y;
}
```

**Comment by YairHalberstadt** [comment, marked as answer]
> For hi-perf computing it's not always helpful to have getters and setters

I'm interested what this case is. Usually the jit will inline pretty much all autoprops.

**Reply by msedi** [reply]
@YairHalberstadt: That's a valid argument! I haven't really tested this nowadays - lets say it has been a bottleneck years ago and I didn't follow it. I will do.

The question nevertheless is will there be some init for fields?

**Reply by YairHalberstadt** [reply]
For example these two methods produce identical jit:

https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxNMAfAAgJgIwCwAUFgAwAEWeAdAEoCuAdjAJZhzUDCA9mAA4sANogDKiAG4sAxnCgBuYsSwBmSjnKdyAb2Lk9lVVgAs5ALJ4AFAEptu/ffEQE5PtxbNyAXnKM4Ad3IABTcPLXIAfQAPL3I8JAiATxj1AF8FInt7ABUIAGtZYPcYC1ciq3T7FLs9aoNKE1Mca1sMzL1HZ1KPb18AwtDyAA0YuPIATWTyNNrsvIKQ4q6Yctqq1pr18gBtUzgYAAtuABMASX5BC12D47O+QQB5PlZuRihqADluE8ZBd3cAcysAF1aip6uQcvkoP1ijCXAsrFo1mslKpYAh6FIYEEFsQdJswUUIpEKvpCR5wglSXpydjhmF/nsvAA+YlychQJmeVlRGKOQT0ODstb2WnjbTkRnY7mJdmc6U8pLefmC4XEFJAA=

I don't believe there's any plans right now. It is a reasonable suggestion.

If you can come up with some realistic use cases I would definitely say open a proposal.

It should be fairly straightforward - take the existing `init` property proposal and modify as necessary for fields.

**Reply by msedi** [reply]
@YairHalberstadt: Thanks for verifying this. I will try to open a proposal with some example for this.

**Reply by PathogenDavid** [reply]
If you do make a proposal, you should point out that this would allow taking the address of the field in unsafe scenarios.

[You can convince the JIT to inline enough to take a pointer to the backing field, but it isn't pretty:](https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxNMAfAAgBgARYCMAdAEoCuAdjAJZhzEDCA9mAA40A2iAyogG40AxnCgBuALAAoaVgDMeKlAgAzOAQBMeRngDe0vIYIKsAFjwBZQgAoAlHoNGn/CAjxtmNangC8eSnAA7ngACp7eungA+gAevniESNEAnvFaAL6SUk5OAPS5ACoQANaiYV4w1h4Vtlk5RvlFpVAAktRV4TDEUcm1jvVNom0wITAI1gBk1dTdvXVG6f14S/IE5hYadg7Z9XgubtMw8QHB5RF4ABrxiXgAmml4mUt5hSVlnR018/WNb63th2Itz6OwGf2Go3G1gqACpbABVSjKNTEACCUDO8HGCDgKjwiORDHRpFxnxmwNsIKci1By1pAG0LHAYAALZgAExa7E41iZrI5XLYnAA8mxaMwkcQAHLMNqcLxeADmtgAuisTOZBhiPpj3J1bLoaU4lozmWzOdzeWaBdzReLJTK5QrKMq1bTVmY8FrhtDKHgKngYvZDUsTXzzYKeeGbUK7TQJVBpbLKPLKErVeq1l7wdRIb6YDDA8GaTTZApYAhyEIjpjpPp3QoA7FvsZ/d4ei3VgGrpFFczfAA+aIxMR4KD9nxD2LxFycchwUdGoxd7z3XsTqfJUfjo6TlIziBzhePaTpIA==)

```csharp
TakesIntPtr((int*)Unsafe.AsPointer(ref Unsafe.AsRef(point.Y)));
```

It also [doesn't always work](https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxNMAfAAgBgARYCMAdAEoCuAdjAJZhzEDCA9mAA40A2iAyogG40AxnCgBuALAAoaVgDMeKlAgAzOAQBMeRtIDe0vIYIKsAFjwBZQgH1ScFQAoalPAAVmzmHjYfqASgMjfSkjULwAenCAFQgAa1EASWoHH09iawBPP0kQsMMVGgAPOAATPAd3TwAqb18YVzwAXjwAMlT/QLzgvLyY+Kgk+pgEBza61wBaAD5M7M6wgF95pdzDefkCcwsNW3snF0rqWs8A1bxunsi+xOT2mGIATTmz0OuB6ldhh32YKr8AVUoyjUxAAglBDvARgh7HhAcCGOC7I47o8/OicqEVqF5gBtCxwGAAC2YJQS7E4DgJxNJ5LYnAA8mxaMwgcQAHLMJKcZzOADmfgAuusTOY3pCKnVjv5dNijHjqSSyRSqYSlXTGcyaKyoByuZQeZR+UKRZs8G9Bvs8J48IU/Oc5Wszvi1bSVYq3fSmSy2ZzubzKALhWcNmZzXEbkMRtVbfbZdIVrIFLAEOQhF5IXpTTbrIVMUYNjmMvnDIWjgANc54PmEppTPC5sR4KC1xr13NNPD8CCcchwJuO4zWo4PKs1rxthvF5ut9sZTvd3v9vArBZAA=). (`M2_Ref` is copying the struct onto the stack in both the `TakesInt` and `TakesIntPtr` cases.)
